File: Java/백준/Silver/1325. 효율적인 해킹/효율적인 해킹.java
import java.io.*;
import java.util.*;

public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static StringTokenizer st;
    static int N, M, A, B, cnt;
    static ArrayList<ArrayList<Integer>> vertex;
    static ArrayList<Integer> lst;
    static boolean[] visited;

    public static void main(String[] args) throws IOException {

        input();

        solve();

        br.close();
        bw.close();
    }

    private static void dfs(int curr) {

        for (Integer nxt : vertex.get(curr)) {
            if (!visited[nxt]) {
                visited[nxt] = true;
                cnt++;
                lst.add(nxt);
                dfs(nxt);
            }
        }

    }

    private static void solve() throws IOException {

        int MAX = 0;

        ArrayList<Integer> ans = new ArrayList<>();
        for (int i = 1; i <= N; i++) {
            visited = new boolean[N + 1];
            lst = new ArrayList<>();
            cnt = 1;
            visited[i] = true;
            dfs(i);

            if (MAX < cnt) {
                ans.clear();
                ans.add(i);
                MAX = cnt;
            } else if (MAX == cnt) ans.add(i);
        }

        Collections.sort(ans);
        for (Integer idx : ans)
            bw.write(idx + " ");
    }


    private static void input() throws IOException {

        st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        vertex = new ArrayList<>();
        for (int i = 0; i <= N; i++)
            vertex.add(new ArrayList<Integer>());

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            A = Integer.parseInt(st.nextToken());
            B = Integer.parseInt(st.nextToken());

            // A -> B 단방향 관계
            vertex.get(B).add(A);
        }
    }
}

File: Python/백준/Silver/1325. 효율적인 해킹/효율적인 해킹.py
from collections import deque


def bfs(x):
    dq = deque([x])
    visited = [False] * (n + 1)
    visited[x] = True

    cnt = 1
    while dq:
        curr = dq.popleft()
        for nxt in lst[curr]:
            if not visited[nxt]:
                visited[nxt] = True
                dq.append(nxt)
                cnt += 1

    return cnt


n, m = map(int, input().split())
lst = [[] for _ in range(n + 1)]

ans = []
for _ in range(m):
    a, b = map(int, input().split())
    # a가 b를 신뢰하면 b -> a 접근이 가능하다.
    lst[b].append(a)

MAX = -1
for i in range(1, n + 1):
    cnt = bfs(i)
    if MAX < cnt:
        ans = [i]
        MAX = cnt
    elif MAX == cnt:
        ans.append(i)

print(*ans)
